### ISSUE TITLE
Cryptic error on Map/Set/WeakMap/WeakSet

### ISSUE THREAD

**Issue by thany** [Microsoft/TypeScript#29751]
**TypeScript Version:**  3.4.0-dev.20190202

**Search Terms:** Map, WeakMap, Set

**Code**
```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
const instances = new WeakMap<HTMLElement, string>();
```
Feel free to replace `WeakMap` by `Map`, or `WeakSet`/`Set` (but with one type). The problem will persist.

**Expected behavior:**
No error. This code is perfectly fine. Without the types, it would be valid javascript.

**Actual behavior:**
```
test.ts:1:23 - error TS2693: 'WeakMap' only refers to a type, but is being used as a value here.

1 const instances = new WeakMap<HTMLElement, string>();
                        ~~~~~~~
```
The code does compile & function absolutely fine. As expected, because there's nothing wrong with it.

**Playground Link:** 
[Playground doesn't show the error, but here it is anyway](https://www.typescriptlang.org/play/#src=const%20instances%20%3D%20new%20WeakMap%3CHTMLElement%2C%20string%3E()%3B)

**Related Issues:** 
None that I coud find.

**Comment by dragomirtitian** [comment]
The error is due to the fact that only the type for `WeakMap` is defined not the constructor for it. The constructor is defined in `lib.es2015.collection.d.ts`. The type may be coming from a library (for example I have `lodash` installed and it defines the type regardless of target version, probably in order to make the definitions usable across any target/lib combination)

Solution: 
If you use `target: 'es2015'` or above the error should go away. 

If you want to target a lower version, you can still use the `lib` option to include the definitions for `WeakMap`:
```json
{
    "compilerOptions": {
        "target": "es5",
        "lib": [
            "dom",
            "scripthost",
            "es5",
            "es2015.collection"
        ]
    }
}
```

**Comment by DanielRosenwasser** [comment]
When common types exist (here as forward declarations) but not their analogous constructors, we can potentially provide a more specialized declaration.

**Comment by DanielRosenwasser** [comment]
We already do this for `Promise` and `Symbol` 🤦‍♂️

**Comment by thany** [comment]
> Solution:
> If you use `target: 'es2015'` or above the error should go away.
> 
> If you want to target a lower version, you can still use the `lib` option to include the definitions for `WeakMap`:

Unfortunately our project is a little bit more complex than that.

However, the problem remains that Typescript is assuming the parser dictates which types may be available. In practise, this may be the case, but TS doesn't know about any polyfills we're using.

In short: it shouldn't blindly assume the absence of a type based on ES5/6/2017/etc selection.
